cmd/soroban-rpc/internal/db/event.go

```diff
@@ -0,0 +1,99 @@
+package db
+
+import (
+	sq "github.com/Masterminds/squirrel"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/support/db"
+	"github.com/stellar/go/support/errors"
+	"github.com/stellar/go/support/log"
+	"github.com/stellar/go/xdr"
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/events"
+	"io"
+)
+
+const eventTableName = "events"
+
+type EventWriter interface {
+	InsertEvents(lcm xdr.LedgerCloseMeta) error
+}
+
+type EventReader interface {
+	GetEvents(lcm xdr.LedgerCloseMeta) error
+}
+
+type eventHandler struct {
+	log                       *log.Entry
+	db                        db.SessionInterface
+	stmtCache                 *sq.StmtCache
+	passphrase                string
+	ingestMetric, countMetric prometheus.Observer
+}
+
+func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
+	txCount := lcm.CountTransactions()
+
+	if eventHandler.stmtCache == nil {
+		return errors.New("EventWriter incorrectly initialized without stmtCache")
+	} else if txCount == 0 {
+		return nil
+	}
+
+	var txReader *ingest.LedgerTransactionReader
+	txReader, err := ingest.NewLedgerTransactionReaderFromLedgerCloseMeta(eventHandler.passphrase, lcm)
+	if err != nil {
+		return errors.Wrapf(err,
+			"failed to open transaction reader for ledger %d",
+			lcm.LedgerSequence())
+	}
+	defer func() {
+		closeErr := txReader.Close()
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	for {
+		var tx ingest.LedgerTransaction
+		tx, err = txReader.Read()
+		if err == io.EOF {
+			err = nil
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if !tx.Result.Successful() {
+			continue
+		}
+
+		txEvents, err := tx.GetDiagnosticEvents()
```

Comment:
This shadows the `err`above, so the error handling in defer above won't do what you expect. You may want to use a named return parameter to make it cleaner.